src/videoencodingmonitor/VideoEncodingMonitor.ts

```diff
@@ -0,0 +1,173 @@
+            isHardwareEncoder = value as boolean;
+            break;
+          case 'videoUpstreamCpuQualityLimitationDurationPerSecond':
+            cpuLimitationDuration = Math.trunc(Number(value));
+            break;
+        }
+      }
+    }
+
+    const cpuUsageIsHigh =
+      !isHardwareEncoder &&
+      (totalEncodingTimeInMs >= VideoEncodingMonitor.HIGH_ENCODE_CPU_MS_THRESHOLD ||
+        cpuLimitationDuration > 0);
+    if (cpuUsageIsHigh) {
+      this.continuousHighEncodeCpuCnt++;
+      if (
+        this.continuousHighEncodeCpuCnt > VideoEncodingMonitor.CONTINUOUS_HIGH_ENCODE_CPU_THRESHOLD
+      ) {
+        this.degradeVideoCodec();
+        this.context.statsCollector.videoCodecDegradationHighEncodeCpuDidReceive();
+        this.continuousHighEncodeCpuCnt = 0;
+      }
+    } else {
+      this.continuousHighEncodeCpuCnt = 0;
+    }
+  }
+
+  /**
+   * Check video encoding metrics for hardware video encoder failures and
+   * degrade video codec when necessary.
+   */
+  private checkHwEncodeFailure(videoMetricReport: { [id: string]: { [id: string]: {} } }): void {
+    let videoInputFps: number = 0;
+    let videoEncodeFps: number = 0;
+    let isHardwareEncoder: boolean = false;
+
+    const attendeeId = this.context.meetingSessionConfiguration.credentials.attendeeId;
+
+    const metricData: { [id: string]: { [id: string]: {} } } = videoMetricReport[attendeeId];
+    const streams = metricData ? Object.keys(metricData) : [];
+    if (streams.length === 0) {
+      return;
+    }
+
+    for (const ssrc of streams) {
+      for (const [metricName, value] of Object.entries(metricData[ssrc])) {
+        switch (metricName) {
+          case 'videoUpstreamFramesInputPerSecond':
+            videoInputFps = Math.trunc(Number(value));
+            break;
+          case 'videoUpstreamFramesEncodedPerSecond':
+            videoEncodeFps = Math.trunc(Number(value));
+            break;
+          case 'videoUpstreamEncoderImplementation':
+            isHardwareEncoder = value as boolean;
+            break;
+        }
+      }
+    }
+
+    const hwEncoderFailed = isHardwareEncoder && videoEncodeFps === 0 && videoInputFps > 0;
+    if (hwEncoderFailed) {
+      this.continuousHwEncodeFailureCnt++;
+      if (
+        this.continuousHwEncodeFailureCnt >
+        VideoEncodingMonitor.CONTINUOUS_HW_ENCODE_FAILURE_THRESHOLD
+      ) {
+        this.degradeVideoCodec();
+        this.context.statsCollector.videoCodecDegradationHwEncodeFailureDidReceive();
+        this.continuousHwEncodeFailureCnt = 0;
+      }
+    } else {
+      this.continuousHwEncodeFailureCnt = 0;
+    }
+    return;
+  }
+
+  /**
+   * Degrade video codec to an alternative codec
+   */
+  private degradeVideoCodec(): void {
+    // Degrade video codec if there are other codec options and current codec is not H264 CBP or VP8
+    if (
+      this.context.meetingSupportedVideoSendCodecPreferences !== undefined &&
```

Comment:
Probably out of scope for this, but I feel like we should probably try to collect all the logic related to these arrays into a single unit testable class :)